### DISCUSSION TITLE
ARM64 Cross compilation IPV6 - Runtime Error

### DISCUSSION THREAD

**Discussion by SagarGowda89** [strongswan/strongswan#918] [Q&A]
Hi 

I have cross compiled strongswan 5.8 for arm64. For IPV4 conenction established and esp packets exchange working fine. But for IPV6 connection established but No esp packets. i can see below error

Mar 20 22:15:15 00[KNL] received netlink error: Operation not supported (95)
Mar 20 22:15:15 00[KNL] unable to create IPv4 routing table rule
Mar 20 22:15:15 00[KNL] received netlink error: Operation not supported (95)

and

Mar 20 22:15:15 06[KNL] <conn-1|1>   using replay window of 0 packets
Mar 20 22:15:15 06[KNL] <conn-1|1> received netlink error: Protocol not supported (93)
Mar 20 22:15:15 06[KNL] <conn-1|1> unable to add SAD entry with SPI c2b399d8 (FAILED)
Mar 20 22:15:15 06[IKE] <conn-1|1> unable to install inbound and outbound IPsec SA (SAD) in kernel
Mar 20 22:15:15 06[IKE] <conn-1|1> failed to establish CHILD_SA, keeping IKE_SA
Mar 20 22:15:15 06[IKE] <conn-1|1> reinitiating already active tasks

can i know how to fix this issue

**Comment by tobiasbrunner** [comment]
Make sure your kernel has the [required modules](https://wiki.strongswan.org/projects/strongswan/wiki/KernelModules) active.

**Reply by SagarGowda89** [reply]
i can see below plugins.

 strongSwan will be built with the following plugins
-----------------------------------------------------
libstrongswan: aes des rc2 sha2 sha1 md5 mgf1 random nonce x509 revocation constraints pubkey pkcs1 pkcs7 pkcs8 pkcs12 pgp dnskey sshkey pem openssl fips-prf gmp curve25519 xcbc cmac hmac drbg
libcharon:     attr kernel-netlink resolve socket-default stroke vici updown eap-identity eap-sim eap-sim-file eap-aka xauth-generic counters
libtnccs:
libtpmtss:


I dont see below module. 

ah4
ah6
esp4
esp6
xfrm4_tunnel
xfrm6_tunnel
xfrm_user
ip_tunnel
tunnel
tunnel6
xfrm4_mode_tunnel
xfrm6_mode_tunnel



can i know how to check/load these modules?

**Reply by tobiasbrunner** [reply]
Plugins loaded by the IKE daemon, which runs in userland, have nothing to do with Linux kernel modules. Refer to the page I linked above for a way to check for the latter.

**Reply by SagarGowda89** [reply]
Below are the module i have enabled. still issue is not resolved

CONFIG_XFRM_USER=y
CONFIG_NET_KEY=y
CONFIG_INET=y
CONFIG_IP_ADVANCED_ROUTER=y
CONFIG_INET_AH=y
CONFIG_INET_ESP=y
CONFIG_INET_IPCOMP=y
CONFIG_INET_XFRM_MODE_TRANSPORT=y
CONFIG_INET_XFRM_MODE_TUNNEL=y
CONFIG_INET_XFRM_MODE_BEET=y
CONFIG_IPV6=y
CONFIG_INET6_AH=y
CONFIG_INET6_ESP=y
CONFIG_INET6_IPCOMP=y
CONFIG_INET6_XFRM_MODE_TRANSPORT=y
CONFIG_INET6_XFRM_MODE_TUNNEL=y
CONFIG_INET6_XFRM_MODE_BEET=y
CONFIG_IPV6_MULTIPLE_TABLES=y
CONFIG_NETFILTER=y
CONFIG_NETFILTER_XTABLES=y
CONFIG_NETFILTER_XT_MATCH_POLICY=y



Error in charon.log

Mar 17 21:43:29 10[KNL] <conn-1|1>   HW offload: no
Mar 17 21:43:29 10[KNL] <conn-1|1> received netlink error: Protocol not supported (93)
Mar 17 21:43:29 10[KNL] <conn-1|1> unable to add SAD entry with SPI c9fb0bc3 (FAILED)
Mar 17 21:43:29 10[CHD] <conn-1|1> adding outbound ESP SA
Mar 17 21:43:29 10[CHD] <conn-1|1>   SPI 0xc49762b8, src fd5d:d50a:8c17:2010::a02 dst fd5d:d50a:8c17:2010::a01
Mar 17 21:43:29 10[KNL] <conn-1|1> adding SAD entry with SPI c49762b8 and reqid {1}
Mar 17 21:43:29 10[KNL] <conn-1|1>   using encryption algorithm AES_CBC with key size 128
Mar 17 21:43:29 10[KNL] <conn-1|1>   using integrity algorithm HMAC_SHA1_96 with key size 160
Mar 17 21:43:29 10[KNL] <conn-1|1>   using replay window of 0 packets
Mar 17 21:43:29 10[KNL] <conn-1|1>   HW offload: no
Mar 17 21:43:29 10[KNL] <conn-1|1> received netlink error: Protocol not supported (93)
Mar 17 21:43:29 10[KNL] <conn-1|1> unable to add SAD entry with SPI c49762b8 (FAILED)
Mar 17 21:43:29 10[IKE] <conn-1|1> unable to install inbound and outbound IPsec SA (SAD) in kernel
Mar 17 21:43:29 10[IKE] <conn-1|1> failed to establish CHILD_SA, keeping IKE_SA
Mar 17 21:43:29 10[IKE] <conn-1|1> reinitiating already active tasks

**Reply by tobiasbrunner** [reply]
> CONFIG_INET6_XFRM_MODE_TRANSPORT=m
> CONFIG_INET6_XFRM_MODE_TUNNEL=m
> CONFIG_INET6_XFRM_MODE_BEET=m

Presumably, these are not loaded. Either don't compile them as modules (like the others) or load them manually (e.g. `modprobe xfrm6_mode_tunnel` or `modprobe xfrm6_mode_transport`).

**Reply by SagarGowda89** [reply]
yes its done. no changes

Module                                  Size  Used by
xfrm6_mode_beet                 16384  0
xfrm6_mode_transport         16384  0
xt_addrtype                          16384  2
xt_conntrack                         16384  1
crc32_ce                               16384  0
xfrm6_mode_tunnel             16384  0
pfe                                       98304  0
lm90                                    28672  0
ina2xx                                 16384  0

**Reply by SagarGowda89** [reply]
root@localhost:# ls kernel/net/ipv6
netfilter  xfrm6_mode_beet.ko  xfrm6_mode_transport.ko  xfrm6_mode_tunnel.ko
root@localhost:#ls kernel/net/ipv6/netfilter/ip6
ip6_tables.ko       ip6t_MASQUERADE.ko  ip6t_REJECT.ko      ip6table_filter.ko  ip6table_mangle.ko  ip6table_nat.ko



here i cannot see "esp6.ko"   can this will create issue ?

Mar 17 23:37:52 09[KNL] <conn-1|1>   using replay window of 32 packets
Mar 17 23:37:52 09[KNL] <conn-1|1>   HW offload: no
Mar 17 23:37:52 09[KNL] <conn-1|1> received netlink error: Protocol not supported (93)
Mar 17 23:37:52 09[KNL] <conn-1|1> unable to add SAD entry with SPI c9c9c2f5 (FAILED)
Mar 17 23:37:52 09[CHD] <conn-1|1> adding outbound ESP SA
Mar 17 23:37:52 09[CHD] <conn-1|1>   SPI 0xc9eda9aa, src fd5d:d50a:8c17:2010::a02 dst fd5d:d50a:8c17:2010::a01
Mar 17 23:37:52 09[KNL] <conn-1|1> adding SAD entry with SPI c9eda9aa and reqid {1}
Mar 17 23:37:52 09[KNL] <conn-1|1>   using encryption algorithm AES_CBC with key size 128
Mar 17 23:37:52 09[KNL] <conn-1|1>   using integrity algorithm HMAC_SHA1_96 with key size 160
Mar 17 23:37:52 09[KNL] <conn-1|1>   using replay window of 0 packets
Mar 17 23:37:52 09[KNL] <conn-1|1>   HW offload: no

**Reply by tobiasbrunner** [reply]
If it's not built as module (`=y`), you won't see a `*.ko` file, the code is just directly included in the kernel image. Maybe try to build all of these that way.

**Reply by SagarGowda89** [reply]
I do not have permission to rebuild kernel.  when i try to insert module i am getting below error.

root@localhost:~# modprobe ah
modprobe: FATAL: Module ah not found in directory /lib/modules/4.19
root@localhost:~# modprobe esp6
modprobe: FATAL: Module esp6 not found in directory /lib/modules/4.19



i have checked required module present of not then i got below data. i have tried to change it to enabled ("y") like mentioned in previous message but no use but i am not rebuilding so i have reverted changes to old as shown below

CONFIG_XFRM_USER=y
CONFIG_NET_KEY=y
CONFIG_INET=y
 //# CONFIG_IP_ADVANCED_ROUTER is not set
 //# CONFIG_INET_AH is not set
CONFIG_INET_ESP=y
 //# CONFIG_INET_IPCOMP is not set
CONFIG_INET_XFRM_MODE_TRANSPORT=y
CONFIG_INET_XFRM_MODE_TUNNEL=y
CONFIG_INET_XFRM_MODE_BEET=y
CONFIG_IPV6=y
 //# CONFIG_INET6_AH is not set
 //# CONFIG_INET6_ESP is not set
 //# CONFIG_INET6_IPCOMP is not set
CONFIG_INET6_XFRM_MODE_TRANSPORT=m
CONFIG_INET6_XFRM_MODE_TUNNEL=m
CONFIG_INET6_XFRM_MODE_BEET=m
 //# CONFIG_IPV6_MULTIPLE_TABLES is not set
CONFIG_NETFILTER=y
CONFIG_NETFILTER_XTABLES=y
 //# CONFIG_NETFILTER_XT_MATCH_POLICY is not set

**Reply by tobiasbrunner** [reply]
> //# CONFIG_INET6_ESP is not set

It obviously won't work without this. So you have to rebuild your kernel somehow.
